File: taiwan_stock_app/frontend/lib/models/ai_config.dart
/// AI Config Models — BYOK 設定

class AIModelInfo {
  final String id;
  final String label;

  AIModelInfo({required this.id, required this.label});

  factory AIModelInfo.fromJson(Map<String, dynamic> json) {
    return AIModelInfo(
      id: json['id'] as String,
      label: json['label'] as String,
    );
  }
}

class AIProviderInfo {
  final String id;
  final String label;
  final List<AIModelInfo> models;

  AIProviderInfo({required this.id, required this.label, required this.models});

  factory AIProviderInfo.fromJson(Map<String, dynamic> json) {
    return AIProviderInfo(
      id: json['id'] as String,
      label: json['label'] as String,
      models: (json['models'] as List)
          .map((m) => AIModelInfo.fromJson(m as Map<String, dynamic>))
          .toList(),
    );
  }
}

class AIConfig {
  final String? provider;
  final String? model;
  final bool hasApiKey;
  final String? providerLabel;

  AIConfig({
    this.provider,
    this.model,
    this.hasApiKey = false,
    this.providerLabel,
  });

  factory AIConfig.fromJson(Map<String, dynamic> json) {
    return AIConfig(
      provider: json['provider'] as String?,
      model: json['model'] as String?,
      hasApiKey: json['has_api_key'] as bool? ?? false,
      providerLabel: json['provider_label'] as String?,
    );
  }
}

File: taiwan_stock_app/backend/app/schemas/ai_config.py
"""
AI Config Schemas — BYOK 設定的請求/回應模型
"""

from pydantic import BaseModel, field_validator
from typing import List, Optional


# 可用的 provider + model 清單
AI_PROVIDERS = {
    "gemini": {
        "label": "Google Gemini",
        "models": ["gemini-2.0-flash", "gemini-2.5-pro"],
    },
    "openai": {
        "label": "OpenAI",
        "models": ["gpt-4o-mini", "gpt-4o", "gpt-4.1-mini", "gpt-4.1"],
    },
    "groq": {
        "label": "Groq",
        "models": ["llama-3.3-70b-versatile", "llama-4-maverick-17b-128e"],
    },
}

VALID_PROVIDERS = set(AI_PROVIDERS.keys())
VALID_MODELS = {m for p in AI_PROVIDERS.values() for m in p["models"]}


class AIModelInfo(BaseModel):
    id: str
    label: str


class AIProviderInfo(BaseModel):
    id: str
    label: str
    models: List[AIModelInfo]


class AIProvidersResponse(BaseModel):
    providers: List[AIProviderInfo]


class AIConfigSave(BaseModel):
    provider: str
    model: str
    api_key: str

    @field_validator("provider")
    @classmethod
    def validate_provider(cls, v: str) -> str:
        v = v.lower().strip()
        if v not in VALID_PROVIDERS:
            raise ValueError(f"不支援的 provider: {v}，可選: {', '.join(VALID_PROVIDERS)}")
        return v

    @field_validator("model")
    @classmethod
    def validate_model(cls, v: str) -> str:
        v = v.strip()
        if v not in VALID_MODELS:
            raise ValueError(f"不支援的 model: {v}")
        return v

    @field_validator("api_key")
    @classmethod
    def validate_api_key(cls, v: str) -> str:
        v = v.strip()
        if len(v) < 10:
            raise ValueError("API Key 長度不足")
        return v


class AIConfigResponse(BaseModel):
    provider: Optional[str] = None
    model: Optional[str] = None
    has_api_key: bool = False
    provider_label: Optional[str] = None


class AIConfigTestRequest(BaseModel):
    provider: str
    model: str
    api_key: str

    @field_validator("provider")
    @classmethod
    def validate_provider(cls, v: str) -> str:
        v = v.lower().strip()
        if v not in VALID_PROVIDERS:
            raise ValueError(f"不支援的 provider: {v}")
        return v
